File: windows/PngCut.Tests/SharedContractBehaviorTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using NUnit.Framework;
using PngCut.Core.Models;
using PngCut.Core.Services;

namespace PngCut.Tests;

public class SharedContractBehaviorTests
{
    private string _temporaryRoot = null!;

    [SetUp]
    public void SetUp()
    {
        _temporaryRoot = Path.Combine(Path.GetTempPath(), "PngCut-SharedContractBehaviorTests-" + Guid.NewGuid());
        Directory.CreateDirectory(_temporaryRoot);
    }

    [TearDown]
    public void TearDown()
    {
        if (Directory.Exists(_temporaryRoot))
        {
            Directory.Delete(_temporaryRoot, true);
        }
    }

    [Test]
    public void Adjacent_case_normalizes_the_output_extension()
    {
        var testCase = OutputCase("adjacent-normalizes-extension");
        var source = CreateFile(testCase.Source);

        var prepared = OutputPolicy.Prepare(source, OutputMode.Adjacent);

        Assert.That(Relative(prepared.FinalPath), Is.EqualTo(testCase.Expected));
    }

    [Test]
    public void Folder_case_preserves_the_relative_source_name()
    {
        var testCase = OutputCase("folder-preserves-relative-source-name");
        var source = CreateFile(testCase.Source);

        var prepared = OutputPolicy.Prepare(
            source,
            OutputMode.Adjacent,
            selectedFolderRoot: LogicalPath(testCase.SelectedFolder!));

        Assert.That(Relative(prepared.FinalPath), Is.EqualTo(testCase.Expected));
    }

    [Test]
    public void Adjacent_collision_uses_the_first_numbered_suffix()
    {
        var testCase = OutputCase("adjacent-collision-uses-two");
        var source = CreateFile(testCase.Source);
        foreach (var existingOutput in testCase.ExistingOutputs ?? new List<string>())
        {
            CreateFile(existingOutput);
        }

        var prepared = OutputPolicy.Prepare(source, OutputMode.Adjacent);

        Assert.That(Relative(prepared.FinalPath), Is.EqualTo(testCase.Expected));
    }

    [Test]
    public void Overwrite_case_keeps_the_source_and_allows_replacement()
    {
        var testCase = OutputCase("overwrite-keeps-source");
        var source = CreateFile(testCase.Source);

        var prepared = OutputPolicy.Prepare(source, OutputMode.Overwrite);

        Assert.That(Relative(prepared.FinalPath), Is.EqualTo(testCase.Expected));
        Assert.That(prepared.AllowsReplacingExistingFile, Is.True);
    }

    [Test]
    public void Custom_case_rejects_a_missing_directory_with_the_contract_code()
    {
        var testCase = OutputCase("custom-requires-existing-directory");
        var source = CreateFile(testCase.Source);

        Assert.That(
            () => OutputPolicy.Prepare(source, OutputMode.Custom, LogicalPath(testCase.CustomDirectory!)),
            Throws.TypeOf<ArgumentException>());
        Assert.That(testCase.ExpectedError, Is.EqualTo("output_policy_invalid"));
    }

    [Test]
    public void Windows_unc_share_root_case_matches_the_shared_contract()
    {
        var testCase = OutputCase("windows-unc-share-root");

        var finalPath = OutputPolicy.ForFolderImport(testCase.SelectedFolder!, testCase.Source);

        Assert.That(finalPath, Is.EqualTo(testCase.Expected));
    }

    [Test]
    public void Discovery_cases_match_the_shared_contract()
    {
        foreach (var testCase in DiscoveryContract().Cases.Where(testCase => testCase.Platforms.Contains("all")))
        {
            ResetTemporaryRoot();
            foreach (var path in testCase.Files)
            {
                CreateFile(path);
            }

            var result = FileDiscovery.Discover(
                testCase.SelectedPaths.Select(LogicalPath),
                CompressionMode.Lossless);

            Assert.That(
                result.Tasks.Select(task => Relative(task.SourcePath)).OrderBy(path => path),
                Is.EqualTo(testCase.ExpectedImages.OrderBy(path => path)),
                testCase.Id);
            Assert.That(result.SkippedRegularFileCount, Is.EqualTo(testCase.ExpectedSkippedRegularFiles), testCase.Id);
        }
    }

    private OutputPolicyContractCase OutputCase(string id) =>
        OutputPolicyContract().Cases.Single(testCase => testCase.Id == id);

    private OutputPolicyContract OutputPolicyContract() =>
        ReadContract<OutputPolicyContract>("output-policy-cases.json");

    private DiscoveryContract DiscoveryContract() =>
        ReadContract<DiscoveryContract>("discovery-cases.json");

    private static T ReadContract<T>(string name)
    {
        var json = File.ReadAllText(Path.Combine(TestContext.CurrentContext.TestDirectory, "Contracts", name));
        return new System.Web.Script.Serialization.JavaScriptSerializer().Deserialize<T>(json);
    }

    private string CreateFile(string logicalPath)
    {
        var path = LogicalPath(logicalPath);
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllText(path, "fixture");
        return path;
    }

    private string LogicalPath(string logicalPath) =>
        Path.Combine(_temporaryRoot, logicalPath.Replace('/', Path.DirectorySeparatorChar));

    private void ResetTemporaryRoot()
    {
        Directory.Delete(_temporaryRoot, true);
        Directory.CreateDirectory(_temporaryRoot);
    }

    private string Relative(string path) =>
        path.Substring(_temporaryRoot.Length)
            .TrimStart(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
            .Replace(Path.DirectorySeparatorChar, '/');

    private sealed class OutputPolicyContract
    {
        public int Version { get; set; }
        public List<OutputPolicyContractCase> Cases { get; set; } = new();
    }

    private sealed class OutputPolicyContractCase
    {
        public string Id { get; set; } = null!;
        public List<string> Platforms { get; set; } = new();
        public string Policy { get; set; } = null!;
        public string Source { get; set; } = null!;
        public string? SelectedFolder { get; set; }
        public string? CustomDirectory { get; set; }
        public List<string>? ExistingOutputs { get; set; }
        public string? Expected { get; set; }
        public string? ExpectedError { get; set; }
    }

    private sealed class DiscoveryContract
    {
        public int Version { get; set; }
        public List<DiscoveryContractCase> Cases { get; set; } = new();
    }

    private sealed class DiscoveryContractCase
    {
        public string Id { get; set; } = null!;
        public List<string> Platforms { get; set; } = new();
        public List<string> SelectedPaths { get; set; } = new();
        public List<string> Files { get; set; } = new();
        public List<string> ExpectedImages { get; set; } = new();
        public int ExpectedSkippedRegularFiles { get; set; }
    }
}

File: macos/PngCutTests/SharedContractBehaviorTests.swift
import XCTest
@testable import PngCut

final class SharedContractBehaviorTests: XCTestCase {
    private var temporaryRoot: URL!

    override func setUpWithError() throws {
        temporaryRoot = FileManager.default.temporaryDirectory
            .appendingPathComponent("PngCut-SharedContractBehaviorTests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: temporaryRoot, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: temporaryRoot)
    }

    func testAdjacentCaseNormalizesTheOutputExtension() throws {
        let testCase = try outputCase(id: "adjacent-normalizes-extension")
        let source = try makeFile(at: testCase.source)

        let prepared = try OutputPolicy.adjacent.prepare(source: source)

        XCTAssertEqual(relativePath(of: prepared.finalURL), testCase.expected)
    }

    func testFolderCasePreservesTheRelativeSourceName() throws {
        let testCase = try outputCase(id: "folder-preserves-relative-source-name")
        let source = try makeFile(at: testCase.source)
        let selectedFolder = logicalURL(testCase.selectedFolder!)
        var planner = OutputPlanner(policy: .adjacent, customDirectory: nil, reservedFinalURLs: [])

        let prepared = try planner.prepare(source: source, importedFolderRoot: selectedFolder)

        XCTAssertEqual(relativePath(of: prepared.finalURL), testCase.expected)
    }

    func testAdjacentCollisionUsesTheFirstNumberedSuffix() throws {
        let testCase = try outputCase(id: "adjacent-collision-uses-two")
        let source = try makeFile(at: testCase.source)
        try (testCase.existingOutputs ?? []).forEach { _ = try makeFile(at: $0) }

        let prepared = try OutputPolicy.adjacent.prepare(source: source)

        XCTAssertEqual(relativePath(of: prepared.finalURL), testCase.expected)
    }

    func testOverwriteCaseKeepsTheSourceAndAllowsReplacement() throws {
        let testCase = try outputCase(id: "overwrite-keeps-source")
        let source = try makeFile(at: testCase.source)

        let prepared = try OutputPolicy.overwrite.prepare(source: source)

        XCTAssertEqual(relativePath(of: prepared.finalURL), testCase.expected)
        XCTAssertTrue(prepared.allowsReplacingExistingFile)
    }

    func testCustomCaseRejectsAMissingDirectoryWithTheContractCode() throws {
        let testCase = try outputCase(id: "custom-requires-existing-directory")
        let source = try makeFile(at: testCase.source)

        XCTAssertThrowsError(
            try OutputPolicy.customDirectory.prepare(
                source: source,
                customDirectory: logicalURL(testCase.customDirectory!)
            )
        ) { error in
            XCTAssertEqual(outputPolicyErrorCode(error), testCase.expectedError)
        }
    }

    func testDiscoveryCasesMatchTheSharedContract() throws {
        for testCase in try SharedContractLoader.discoveryContract().cases where testCase.platforms.contains("all") {
            try resetTemporaryRoot()
            try testCase.files.forEach { _ = try makeFile(at: $0) }

            let result = FileDiscovery().discover(urls: testCase.selectedPaths.map(logicalURL))

            XCTAssertEqual(
                result.files.map(relativePath(of:)).sorted(),
                testCase.expectedImages.sorted(),
                testCase.id
            )
            XCTAssertEqual(result.skippedNonImageCount, testCase.expectedSkippedRegularFiles, testCase.id)
        }
    }

    private func outputCase(id: String) throws -> SharedOutputPolicyCase {
        guard let testCase = try SharedContractLoader.outputPolicyContract().cases.first(where: { $0.id == id }) else {
            throw XCTSkip("Missing output policy contract case \(id).")
        }
        return testCase
    }

    private func makeFile(at logicalPath: String) throws -> URL {
        let url = logicalURL(logicalPath)
        try FileManager.default.createDirectory(
            at: url.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )
        try Data("fixture".utf8).write(to: url)
        return url
    }

    private func logicalURL(_ path: String) -> URL {
        temporaryRoot.appendingPathComponent(path)
    }

    private func resetTemporaryRoot() throws {
        try FileManager.default.removeItem(at: temporaryRoot)
        try FileManager.default.createDirectory(at: temporaryRoot, withIntermediateDirectories: true)
    }

    private func relativePath(of url: URL) -> String {
        String(url.standardizedFileURL.path.dropFirst(temporaryRoot.standardizedFileURL.path.count + 1))
    }

    private func outputPolicyErrorCode(_ error: Error) -> String? {
        guard error is OutputPolicyError else {
            return nil
        }
        return "output_policy_invalid"
    }
}
